src/abr/puffer_ttp.cc

```diff
@@ -54,6 +69,52 @@ void PufferTTP::normalize_in_place(size_t i, vector<double> & input)
   }
 }
 
+
+void PufferTTP::calculate_gaussian_values(){
+    double gaussian_coefficient = 1.0/(std_val_*sqrt(2.0*M_PI));
+    std::vector<double> x_vals;
+    int right_pos =(kernel_size_>>1);
+    int left_pos = 0 - right_pos;
+    // In our current case, kernel_size=21, so right_pos = 10, left_pos = -10
+    // To construct the kernel for bluring, we sample the points whose x-axis is [-10,10]
+    for(int i = left_pos; i<=right_pos; i++){
+        x_vals.push_back(i);
+    }
+    for(double & x: x_vals){
+        double exp_idx = -(x-mean_val_)*(x-mean_val_)/(2.0*std_val_*std_val_);
+        double gaussian_val = gaussian_coefficient * exp(exp_idx);
+        gaussian_kernel_vals_.push_back(gaussian_val);
+    }
+}
+
+
+// For future flexibility, this function just blurs one row at a time. 
+void PufferTTP::blur_probability(int horizontal_index, int format_index){
+  std::vector<double> orignial_prob_vals(dis_sending_time_);
+  for(size_t k = 0; k< dis_sending_time_; k++){
+    orignial_prob_vals[k] = sending_time_prob_[horizontal_index][format_index][k];
+  }
+  int right =  (kernel_size_>>1);
+  int left = -right;
+  double sum_prob = 0.0;
+  size_t dim_num = dis_sending_time_ +1;
+  for(size_t k = 0; k < dim_num; k++){
+    double blurred_val = 0.0;
+    for(int j = left; j <= right; j++){
+      int gaussian_index = j + (kernel_size_ >> 1);
+      int covolute_index = (k+j+dim_num)%dim_num;
+      blurred_val += gaussian_kernel_vals_[gaussian_index] * orignial_prob_vals[covolute_index];
+    }
+    sending_time_prob_[horizontal_index][format_index][k] = blurred_val;
+    sum_prob += blurred_val;
+  }
+  //normalized to make the sum of prob as 1.0
+  for(size_t k = 0; k<= dis_sending_time_; k++){
```

Comment:
Are we sure that we don't want to use `k <= dis_sending_time_` too in the two loops of `k` above?